src/Processors/Formats/Impl/AvroRowInputFormat.cpp

```diff
@@ -1258,20 +1287,29 @@ DataTypePtr AvroSchemaReader::avroNodeToDataType(avro::NodePtr node)
         case avro::Type::AVRO_NULL:
             return std::make_shared<DataTypeNothing>();
         case avro::Type::AVRO_UNION:
+        {
             if (node->leaves() == 1)
             {
                 return avroNodeToDataType(node->leafAt(0));
             }
-            else if (
+            if (
                 node->leaves() == 2
                 && (node->leafAt(0)->type() == avro::Type::AVRO_NULL || node->leafAt(1)->type() == avro::Type::AVRO_NULL))
             {
                 int nested_leaf_index = node->leafAt(0)->type() == avro::Type::AVRO_NULL ? 1 : 0;
                 auto nested_type = avroNodeToDataType(node->leafAt(nested_leaf_index));
                 return nested_type->canBeInsideNullable() ? makeNullable(nested_type) : nested_type;
             }
-            /// FIXME Support UNION has more than two datatypes.
-            throw Exception(ErrorCodes::ILLEGAL_COLUMN, "Avro type  UNION is not supported for inserting.");
+            // Map the leaves to the nested data types
+            DataTypes nested_types;
+            nested_types.reserve(node->leaves());
+
+            for (int i = 0; i != static_cast<int>(node->leaves()); ++i)
```

Comment:
int -> size_t

Reply:
The thing is that I need to cast the `size_t` to `int` at some point due to the Avro `NodePtr` API accepting only `int` when using `leafAt` function to access the leaf. But I have changed it to `size_t` anyway and now am statically casting `size_t` to `int` on every iteration as we now use the value to determine the actual index of the potential `null` type.

Now, the `avro::Decoder::decodeUnionIndex()` function returns `size_t`, which we want to compare with the stored index of the `null` type and I consider this to be more performance-sensitive part of the code as the closure runs for every row being deserialized, so I believe it's worth the extra casting cycles during the build of the deserialization logic, which happens just once.

Reply:
Not relevant anymore after resolving https://github.com/ClickHouse/ClickHouse/pull/69712#pullrequestreview-2313179879